File: regions/system/library/typescript/event-bus/__tests__/event-bus.test.ts
import { vi, describe, it, expect, beforeEach } from 'vitest';
import {
  InMemoryEventBus,
  EventBus,
  EventBusError,
} from '../src/index.js';
import type {
  Event,
  DomainEvent,
  EventHandler,
  EventSubscription,
} from '../src/index.js';

// ---------- helpers ----------

function makeEvent(eventType: string): Event {
  return {
    id: 'evt-1',
    eventType,
    aggregateId: 'agg-1',
    occurredAt: new Date(),
    payload: { key: 'value' },
    timestamp: new Date().toISOString(),
  };
}

function makeDomainEvent(eventType: string, aggregateId = 'agg-1'): DomainEvent {
  return { eventType, aggregateId, occurredAt: new Date() };
}

// ---------- Legacy InMemoryEventBus ----------

describe('InMemoryEventBus', () => {
  it('subscribeしたハンドラがpublish時に呼ばれる', async () => {
    const bus = new InMemoryEventBus();
    const handler = vi.fn();
    bus.subscribe('user.created', handler);
    const event = makeEvent('user.created');
    await bus.publish(event);
    expect(handler).toHaveBeenCalledWith(event);
  });

  it('unsubscribeするとハンドラが呼ばれなくなる', async () => {
    const bus = new InMemoryEventBus();
    const handler = vi.fn();
    bus.subscribe('user.created', handler);
    bus.unsubscribe('user.created');
    await bus.publish(makeEvent('user.created'));
    expect(handler).not.toHaveBeenCalled();
  });

  it('異なるイベントタイプのハンドラは呼ばれない', async () => {
    const bus = new InMemoryEventBus();
    const handler = vi.fn();
    bus.subscribe('user.created', handler);
    await bus.publish(makeEvent('order.placed'));
    expect(handler).not.toHaveBeenCalled();
  });

  it('同じイベントタイプに複数のハンドラを登録できる', async () => {
    const bus = new InMemoryEventBus();
    const h1 = vi.fn();
    const h2 = vi.fn();
    bus.subscribe('user.created', h1);
    bus.subscribe('user.created', h2);
    await bus.publish(makeEvent('user.created'));
    expect(h1).toHaveBeenCalledTimes(1);
    expect(h2).toHaveBeenCalledTimes(1);
  });

  it('ハンドラがない場合もpublishがエラーにならない', async () => {
    const bus = new InMemoryEventBus();
    await expect(bus.publish(makeEvent('unknown'))).resolves.toBeUndefined();
  });
});

// ---------- DDD EventBus ----------

describe('EventBus', () => {
  let bus: EventBus;

  beforeEach(() => {
    bus = new EventBus();
  });

  describe('publish / subscribe', () => {
    it('subscribedハンドラがpublish時に呼ばれる', async () => {
      const received: DomainEvent[] = [];
      const handler: EventHandler<DomainEvent> = {
        handle: async (event) => { received.push(event); },
      };
      bus.subscribe('user.created', handler);
      const event = makeDomainEvent('user.created');
      await bus.publish(event);
      expect(received).toHaveLength(1);
      expect(received[0]).toBe(event);
    });

    it('異なるイベントタイプのハンドラは呼ばれない', async () => {
      const handler: EventHandler = {
        handle: vi.fn(),
      };
      bus.subscribe('user.created', handler);
      await bus.publish(makeDomainEvent('order.placed'));
      expect(handler.handle).not.toHaveBeenCalled();
    });

    it('同じイベントタイプに複数のハンドラを登録できる', async () => {
      const h1: EventHandler = { handle: vi.fn() };
      const h2: EventHandler = { handle: vi.fn() };
      bus.subscribe('test', h1);
      bus.subscribe('test', h2);
      await bus.publish(makeDomainEvent('test'));
      expect(h1.handle).toHaveBeenCalledTimes(1);
      expect(h2.handle).toHaveBeenCalledTimes(1);
    });

    it('ハンドラがない場合もpublishがエラーにならない', async () => {
      await expect(bus.publish(makeDomainEvent('unknown'))).resolves.toBeUndefined();
    });
  });

  describe('EventSubscription', () => {
    it('subscribe()がEventSubscriptionを返す', () => {
      const handler: EventHandler = { handle: vi.fn() };
      const sub = bus.subscribe('test', handler);
      expect(sub.eventType).toBe('test');
      expect(typeof sub.unsubscribe).toBe('function');
    });

    it('unsubscribe()でハンドラが解除される', async () => {
      const handler: EventHandler = { handle: vi.fn() };
      const sub = bus.subscribe('test', handler);
      sub.unsubscribe();
      await bus.publish(makeDomainEvent('test'));
      expect(handler.handle).not.toHaveBeenCalled();
    });

    it('特定のハンドラだけをunsubscribeできる', async () => {
      const h1: EventHandler = { handle: vi.fn() };
      const h2: EventHandler = { handle: vi.fn() };
      const sub1 = bus.subscribe('test', h1);
      bus.subscribe('test', h2);
      sub1.unsubscribe();
      await bus.publish(makeDomainEvent('test'));
      expect(h1.handle).not.toHaveBeenCalled();
      expect(h2.handle).toHaveBeenCalledTimes(1);
    });
  });

  describe('EventBusConfig', () => {
    it('デフォルト設定で動作する', async () => {
      const defaultBus = new EventBus();
      const handler: EventHandler = { handle: vi.fn() };
      defaultBus.subscribe('test', handler);
      await defaultBus.publish(makeDomainEvent('test'));
      expect(handler.handle).toHaveBeenCalledTimes(1);
    });

    it('カスタム設定で動作する', async () => {
      const customBus = new EventBus({ bufferSize: 100, handlerTimeoutMs: 1000 });
      const handler: EventHandler = { handle: vi.fn() };
      customBus.subscribe('test', handler);
      await customBus.publish(makeDomainEvent('test'));
      expect(handler.handle).toHaveBeenCalledTimes(1);
    });

    it('handlerTimeoutMsを超えるとHANDLER_FAILEDエラーになる', async () => {
      const slowBus = new EventBus({ handlerTimeoutMs: 50 });
      const handler: EventHandler = {
        handle: () => new Promise((resolve) => setTimeout(resolve, 200)),
      };
      slowBus.subscribe('test', handler);
      await expect(slowBus.publish(makeDomainEvent('test'))).rejects.toThrow(EventBusError);
      await expect(slowBus.publish(makeDomainEvent('test'))).rejects.toMatchObject({
        code: 'HANDLER_FAILED',
      });
    });
  });

  describe('EventBusError', () => {
    it('HANDLER_FAILEDエラーコードでスローされる', async () => {
      const handler: EventHandler = {
        handle: async () => { throw new Error('boom'); },
      };
      bus.subscribe('test', handler);
      try {
        await bus.publish(makeDomainEvent('test'));
        expect.unreachable('should have thrown');
      } catch (err) {
        expect(err).toBeInstanceOf(EventBusError);
        expect((err as EventBusError).code).toBe('HANDLER_FAILED');
      }
    });

    it('CHANNEL_CLOSEDエラーコード: close後のpublish', async () => {
      bus.close();
      try {
        await bus.publish(makeDomainEvent('test'));
        expect.unreachable('should have thrown');
      } catch (err) {
        expect(err).toBeInstanceOf(EventBusError);
        expect((err as EventBusError).code).toBe('CHANNEL_CLOSED');
      }
    });

    it('CHANNEL_CLOSEDエラーコード: close後のsubscribe', () => {
      bus.close();
      const handler: EventHandler = { handle: vi.fn() };
      expect(() => bus.subscribe('test', handler)).toThrow(EventBusError);
      try {
        bus.subscribe('test', handler);
      } catch (err) {
        expect((err as EventBusError).code).toBe('CHANNEL_CLOSED');
      }
    });

    it('EventBusErrorのプロパティが正しい', () => {
      const err = new EventBusError('test message', 'PUBLISH_FAILED');
      expect(err.message).toBe('test message');
      expect(err.code).toBe('PUBLISH_FAILED');
      expect(err.name).toBe('EventBusError');
      expect(err).toBeInstanceOf(Error);
    });
  });

  describe('DomainEvent', () => {
    it('DomainEventのプロパティが正しい', () => {
      const event = makeDomainEvent('user.created', 'user-123');
      expect(event.eventType).toBe('user.created');
      expect(event.aggregateId).toBe('user-123');
      expect(event.occurredAt).toBeInstanceOf(Date);
    });
  });
});

File: regions/system/library/dart/event_bus/test/event_bus_test.dart
import 'package:test/test.dart';

import 'package:k1s0_event_bus/event_bus.dart';

// ---------- Test helpers ----------

Event makeEvent(String type) => Event(
      id: 'evt-1',
      eventType: type,
      payload: {'key': 'value'},
      timestamp: DateTime.now(),
      aggregateId: 'agg-1',
    );

/// Simple DomainEvent implementation for testing.
class TestDomainEvent implements DomainEvent {
  @override
  final String eventType;
  @override
  final String aggregateId;
  @override
  final DateTime occurredAt;

  TestDomainEvent(this.eventType, {this.aggregateId = 'agg-1'})
      : occurredAt = DateTime.now();
}

/// Collecting handler for tests.
class CollectingHandler<T extends DomainEvent> implements EventHandler<T> {
  final List<T> received = [];

  @override
  Future<void> handle(T event) async {
    received.add(event);
  }
}

/// Handler that throws on invocation.
class FailingHandler<T extends DomainEvent> implements EventHandler<T> {
  @override
  Future<void> handle(T event) async {
    throw Exception('boom');
  }
}

/// Handler that delays longer than the timeout.
class SlowHandler<T extends DomainEvent> implements EventHandler<T> {
  final Duration delay;
  SlowHandler(this.delay);

  @override
  Future<void> handle(T event) async {
    await Future.delayed(delay);
  }
}

// ---------- Legacy InMemoryEventBus tests ----------

void main() {
  group('InMemoryEventBus (legacy)', () {
    late InMemoryEventBus bus;

    setUp(() {
      bus = InMemoryEventBus();
    });

    group('subscribe and publish', () {
      test('ハンドラーが一致するイベントを受信すること', () async {
        final received = <Event>[];
        bus.subscribe('user.created', (e) async => received.add(e));
        await bus.publish(makeEvent('user.created'));
        expect(received, hasLength(1));
        expect(received.first.eventType, equals('user.created'));
      });

      test('ハンドラーが一致しないイベントを受信しないこと', () async {
        final received = <Event>[];
        bus.subscribe('user.created', (e) async => received.add(e));
        await bus.publish(makeEvent('user.deleted'));
        expect(received, isEmpty);
      });

      test('同じイベントに複数のハンドラーを登録できること', () async {
        var count = 0;
        bus.subscribe('test', (e) async => count++);
        bus.subscribe('test', (e) async => count++);
        await bus.publish(makeEvent('test'));
        expect(count, equals(2));
      });
    });

    group('unsubscribe', () {
      test('イベントタイプの全ハンドラーが削除されること', () async {
        final received = <Event>[];
        bus.subscribe('test', (e) async => received.add(e));
        bus.unsubscribe('test');
        await bus.publish(makeEvent('test'));
        expect(received, isEmpty);
      });
    });

    group('Event', () {
      test('全フィールドが正しく保存されること', () {
        final event = makeEvent('test');
        expect(event.id, equals('evt-1'));
        expect(event.eventType, equals('test'));
        expect(event.payload, containsPair('key', 'value'));
      });

      test('DomainEventインターフェースを実装していること', () {
        final event = makeEvent('test');
        expect(event, isA<DomainEvent>());
        expect(event.aggregateId, equals('agg-1'));
        expect(event.occurredAt, isA<DateTime>());
      });
    });
  });

  // ---------- DDD EventBus tests ----------

  group('EventBus', () {
    late EventBus bus;

    setUp(() {
      bus = EventBus();
    });

    group('publish / subscribe', () {
      test('サブスクライブされたハンドラーが一致するイベントを受信すること', () async {
        final handler = CollectingHandler<DomainEvent>();
        bus.subscribe('user.created', handler);
        final event = TestDomainEvent('user.created');
        await bus.publish(event);
        expect(handler.received, hasLength(1));
        expect(handler.received.first, same(event));
      });

      test('ハンドラーが一致しないイベントを受信しないこと', () async {
        final handler = CollectingHandler<DomainEvent>();
        bus.subscribe('user.created', handler);
        await bus.publish(TestDomainEvent('order.placed'));
        expect(handler.received, isEmpty);
      });

      test('同じイベントタイプに複数のハンドラーを登録できること', () async {
        final h1 = CollectingHandler<DomainEvent>();
        final h2 = CollectingHandler<DomainEvent>();
        bus.subscribe('test', h1);
        bus.subscribe('test', h2);
        await bus.publish(TestDomainEvent('test'));
        expect(h1.received, hasLength(1));
        expect(h2.received, hasLength(1));
      });

      test('ハンドラーが登録されていない場合にパブリッシュしてもエラーにならないこと', () async {
        await bus.publish(TestDomainEvent('unknown'));
      });
    });

    group('EventSubscription', () {
      test('subscribeがEventSubscriptionを返すこと', () {
        final handler = CollectingHandler<DomainEvent>();
        final sub = bus.subscribe('test', handler);
        expect(sub.eventType, equals('test'));
        expect(sub.isActive, isTrue);
      });

      test('unsubscribeでハンドラーが削除されること', () async {
        final handler = CollectingHandler<DomainEvent>();
        final sub = bus.subscribe('test', handler);
        sub.unsubscribe();
        expect(sub.isActive, isFalse);
        await bus.publish(TestDomainEvent('test'));
        expect(handler.received, isEmpty);
      });

      test('unsubscribeで特定のハンドラーのみ削除されること', () async {
        final h1 = CollectingHandler<DomainEvent>();
        final h2 = CollectingHandler<DomainEvent>();
        final sub1 = bus.subscribe('test', h1);
        bus.subscribe('test', h2);
        sub1.unsubscribe();
        await bus.publish(TestDomainEvent('test'));
        expect(h1.received, isEmpty);
        expect(h2.received, hasLength(1));
      });
    });

    group('EventBusConfig', () {
      test('デフォルト設定で正常に動作すること', () async {
        final defaultBus = EventBus();
        final handler = CollectingHandler<DomainEvent>();
        defaultBus.subscribe('test', handler);
        await defaultBus.publish(TestDomainEvent('test'));
        expect(handler.received, hasLength(1));
      });

      test('カスタム設定で正常に動作すること', () async {
        final customBus =
            EventBus(const EventBusConfig(bufferSize: 100, handlerTimeoutMs: 1000));
        final handler = CollectingHandler<DomainEvent>();
        customBus.subscribe('test', handler);
        await customBus.publish(TestDomainEvent('test'));
        expect(handler.received, hasLength(1));
      });

      test('ハンドラーのタイムアウト時にHANDLER_FAILEDエラーがスローされること', () async {
        final slowBus =
            EventBus(const EventBusConfig(handlerTimeoutMs: 50));
        slowBus.subscribe<DomainEvent>(
          'test',
          SlowHandler(const Duration(milliseconds: 200)),
        );
        expect(
          () => slowBus.publish(TestDomainEvent('test')),
          throwsA(isA<EventBusError>().having(
            (e) => e.code,
            'code',
            EventBusErrorCode.handlerFailed,
          )),
        );
      });
    });

    group('EventBusError', () {
      test('ハンドラーがスローした場合にHANDLER_FAILEDエラーが発生すること', () async {
        bus.subscribe<DomainEvent>('test', FailingHandler());
        expect(
          () => bus.publish(TestDomainEvent('test')),
          throwsA(isA<EventBusError>().having(
            (e) => e.code,
            'code',
            EventBusErrorCode.handlerFailed,
          )),
        );
      });

      test('クローズ後にパブリッシュするとCHANNEL_CLOSEDエラーが発生すること', () {
        bus.close();
        expect(
          () => bus.publish(TestDomainEvent('test')),
          throwsA(isA<EventBusError>().having(
            (e) => e.code,
            'code',
            EventBusErrorCode.channelClosed,
          )),
        );
      });

      test('クローズ後にサブスクライブするとCHANNEL_CLOSEDエラーが発生すること', () {
        bus.close();
        expect(
          () => bus.subscribe('test', CollectingHandler<DomainEvent>()),
          throwsA(isA<EventBusError>().having(
            (e) => e.code,
            'code',
            EventBusErrorCode.channelClosed,
          )),
        );
      });

      test('エラーのプロパティが正しい値を持つこと', () {
        const err = EventBusError(
          'test message',
          EventBusErrorCode.publishFailed,
        );
        expect(err.message, equals('test message'));
        expect(err.code, equals(EventBusErrorCode.publishFailed));
        expect(err.codeString, equals('PUBLISH_FAILED'));
        expect(err.toString(), contains('PUBLISH_FAILED'));
      });

      test('全エラーコードが正しい文字列表現を持つこと', () {
        expect(
          const EventBusError('', EventBusErrorCode.publishFailed).codeString,
          equals('PUBLISH_FAILED'),
        );
        expect(
          const EventBusError('', EventBusErrorCode.handlerFailed).codeString,
          equals('HANDLER_FAILED'),
        );
        expect(
          const EventBusError('', EventBusErrorCode.channelClosed).codeString,
          equals('CHANNEL_CLOSED'),
        );
      });
    });

    group('DomainEvent', () {
      test('TestDomainEventが正しいプロパティを持つこと', () {
        final event = TestDomainEvent('user.created', aggregateId: 'user-123');
        expect(event.eventType, equals('user.created'));
        expect(event.aggregateId, equals('user-123'));
        expect(event.occurredAt, isA<DateTime>());
      });
    });
  });
}
